File: 238-product-of-array-except-self/238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] res = new int[n];
        res[0] = 1;
        
        for(int i = 1; i < n; ++i) {
            res[i] = res[i - 1]*nums[i - 1];
        }
        
        int num = 1;
        for(int i = n - 1; i >= 0; --i) {
            res[i] = res[i]*num;
            num = num*nums[i];
        }
        
        return res;
    }
    
    /* ------------------------------ Brute Force -----------------------------------*/
    /*
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] res = new int[n];
        Arrays.fill(res,1);
        for(int i = 0; i < n; ++i) {
            for(int j = n - 1; j >= 0; --j) {
                if(i == j) {
                    continue;
                } else {
                    res[i] *= nums[j]; 
                }
            }
        }
        return res;
    }
    */
}

File: 238-product-of-array-except-self/238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [1]*n
        
        for i in range(1, n):
            res[i] = res[i - 1]*nums[i - 1]
        
        num = 1
        for i in reversed(range(n)):
            res[i] = res[i]*num
            num = num*nums[i]
        
        return res